### ISSUE TITLE
Windowed + shuffled residency: the ordering, not the window, holds the whole split

### ISSUE THREAD

**Issue by emfdavid** [emfdavid/insitubatch#66]
Windowed residency is not caused by windows, and not by shuffle. It is caused by the two
together, and today's ordering makes the combination as bad as it can be.

## The 2x2

512-sample store, one sample per chunk, 410 train chunks, `block_chunks=8`, `batch_size=8`.
"Live set" is the peak number of chunks that must be co-resident; "lifetime" is how many
blocks a chunk stays resident between its first and last use.

| | no shuffle | shuffle |
|---|--:|--:|
| **no window** `[0]` | 16 chunks (4%) | 16 chunks (4%) |
| **window** `[0, 1]` | 18 chunks (4%) | **229 chunks (56%)** |

Same offsets, same batch size, same block size: shuffle turns an 18-chunk working set into
229. Mean chunk lifetime goes from 1.1 blocks to 18.0, with a maximum of 51 out of 52 --
some chunk is needed in the first block of the epoch and again in the last.

That is why `working_set_bytes` clamps a windowed shuffled pass to the whole train split.
The clamp is roughly honest: measured live sets are 46-56% of the split against a clamp of
100%.

## Why

`block_shuffled_order` permutes at **chunk** granularity, then slices blocks off that
permutation. A windowed anchor `c` reads chunks `c` and `c+lead`, and chunk `c+lead` is
*also* the anchor chunk of another draw -- so two anchors need the same chunk, and the
permutation puts them in uniformly random blocks. Since the driver fetches each chunk once
per epoch (decode-once), that chunk must stay resident across the gap between them.

Residency is therefore set by chunk **lifetime** under the permutation, not by how many
chunks a batch touches. It gets worse as leads multiply, because each additional lead adds
another pair of anchors that must both be alive:

| offsets | live set | % of split | mean lifetime |
|---|--:|--:|--:|
| `[0]` | 16 | 4% | 1.0 |
| `[0, 24]` | 223 | 54% | 16.7 |
| `[0, 24, 240]` | 190 | 46% | 7.7 |
| `[0, 24, 48, 72]` | **397** | **97%** | 27.1 |

A 24/48/72-hour multi-lead training set -- an ordinary forecasting setup -- holds
essentially the entire split resident.

## Exploratory: ordering along the lead chain

The anchors that share a chunk are not random. For a lead stride `s` they are exactly
`c, c+s, c+2s, ...`. Emitting a chain together makes those anchors adjacent, so the chunk's
lifetime collapses -- and because a chain is spread across the whole store, consecutive
emitted samples are still far apart in source order, unlike a contiguous-run ordering.

| offsets | order | live set | mean lifetime | `shuffle_quality` |
|---|---|--:|--:|--:|
| `[0, 24]` | chunk perm (today) | 223 | 16.7 | 0.99 |
| `[0, 24]` | chain perm | **19** | 1.1 | 0.32 |
| `[0, 24, 48, 72]` | chunk perm (today) | 397 | 27.1 | 0.99 |
| `[0, 24, 48, 72]` | chain perm | **25** | 1.3 | 0.32 |

9-16x less residency. What it costs is mixing, and possibly more than the metric says: the
samples a chain groups are exactly one lead apart, which on a weather store is the same
time of day, so they are diurnally correlated in a way `shuffle_quality` does not see.

A first skill check (synthetic advection store, 1536 steps, 24-step horizon, CPU): model
RMSE 0.4837 (chunk) vs 0.4899 (chain) after one epoch, against an identical persistence
RMSE of 0.7249 in both -- so the loader returns the same data, and the skill difference is
about ordering. One epoch and one seed is noise, and a repeated comparison is running.

**Contiguous-run ordering does not work** and was measured before this: it helps only when
the lead is smaller than a run (`[0,1]`: 229 -> 41) and is *worse* at a 24-hour lead
(240 vs 223), while costing the same mixing. Chain ordering follows the stride, which is
why it survives where runs do not.

## What to decide

1. Is 0.32 mixing acceptable for a windowed dataset, given the samples grouped are one
   lead apart? This is an ML question, and the persistence-RMSE harness above is the
   instrument for it.
2. If it is, does chain ordering become automatic when offsets are present (there is no
   sane reason to choose the 97%-residency ordering), or a second shuffle mode? A knob here
   would sit badly with "one obvious way to do it".
3. `[0, 1]` degenerates to sequential order (`shuffle_quality` 0.01), so any rule has to be
   conditional on the lead stride relative to `block_chunks`.

Not a change to make quietly: it alters the order samples arrive in, which is a training
contract, not an implementation detail. Hence an issue rather than a PR.

Related: #65 (bounded read-ahead and the residency floor -- the measurements above use its
`read_ahead_bound` to compute live sets), and the "re-fetch the spill" note in
`working_set_bytes`, which is the other way out of the same problem and trades decode-once
for memory instead of trading mixing.

**Comment by emfdavid** [comment]
Two updates: the chain-ordering experiment finished, and it argues against that direction.

## Chain ordering costs skill

Synthetic advection store (1536 steps, 24-step horizon), 4 epochs, 3 repeats, arms
alternating within each repeat, CPU:

```
order                repeat  model RMSE  persistence  skill %
chunk perm (today)        0      0.4800       0.7249     33.8
chain perm                0      0.4945       0.7249     31.8
chunk perm (today)        1      0.4890       0.7249     32.5
chain perm                1      0.4950       0.7249     31.7
chunk perm (today)        2      0.4839       0.7249     33.2
chain perm                2      0.4968       0.7249     31.5

  chunk perm (today)   skill mean 33.19%   range 32.54-33.79
  chain perm           skill mean 31.65%   range 31.47-31.77
  persistence RMSE distinct values: {0.724868}
```

**~1.5 points of skill, and the arms do not overlap across three repeats.** Persistence RMSE
is identical to six decimals in all six runs, so the loader returned the same data and the
gap is the ordering, not corruption.

So chain ordering buys 9-16x less residency for a measurable training cost. On a synthetic
store, at four epochs -- but it is the wrong direction to spend a real cost on when a
mechanism exists that costs no skill at all.

Scope caveat: synthetic advection has no diurnal structure, which is the specific correlation
I worried about above. A real archive could be worse for chain ordering, not better.

## The examples exercise this today

Not just a hypothetical for large stores. `examples/advection` builds a windowed dataset and
shuffles, so it takes this path on every run:

```
resident 14 chunks, 17 MiB of 68 MiB     <- the 68 MiB budget is the whole-split clamp
```

It is harmless there only because the synthetic store is 16 chunks. The same code against
`era5_c1` sizes the budget to the whole train split -- 4,800 chunks x 4.15 MB, about 20 GB --
and `--source wb2` puts the examples on exactly that footing.

## Proposed: unpin and re-read, with the cache on local NVMe

Rather than reordering, drop the retention requirement that creates the problem. Today a
windowed chunk is pinned from its first use to its last, and under shuffle those are most of
an epoch apart, so residency is the whole split. If a spilled chunk may instead be
**released after each block and re-admitted when a later block needs it**, residency returns
to the two-block floor and ordering is left alone -- no skill cost, because the samples still
arrive in the same order.

What it trades is a re-read. That is why the NVMe guidance matters and is not incidental:
with `cache_dir` on local NVMe the re-read is served from the mmap tier, so it is a local
page-cache read of already-decoded bytes -- not a refetch over the network and not a second
decode. Without a `cache_dir` the same option would re-fetch and re-decode, which is a much
worse trade and probably should not be the default.

Worth noting the current mitigation is already half of this: `cache_dir` makes whole-split
residency *affordable* (slots are file-backed and reclaimable rather than anonymous), which
is why `docs/tuning.md` says to point it at NVMe. The proposal removes the need to size the
budget for the whole split at all.

Design questions this raises, in the order they need answering:

1. **Is it a knob or a rule?** "Retain in RAM unless a `cache_dir` exists, in which case
   release and re-read from it" is a rule with no new argument, and it fails safe: no local
   cache means today's behaviour. A boolean would sit badly with one-obvious-way.
2. **What does `working_set_bytes` return under it?** The whole-split clamp exists precisely
   because retention is unbounded; with re-read it should fall back to the two-block floor,
   and the `describe()` numbers change with it.
3. **How is the re-read counted?** A chunk re-admitted from the local tier is a cache hit,
   not a miss, and the per-epoch INFO line should say so or the hit rate becomes misleading.

If that shape looks right I will open a separate implementation issue for it and leave this
one as the measurement and the ordering post-mortem.

**Comment by emfdavid** [comment]
Prerequisite filed as #68. The per-block release this issue asks for is what breaks the plan/consumer symmetry that has been hiding a defect in `_partition_blocks`: the driver would plan a block from `blocks[i].cids` while the consumer releases `_block_read_keys(order[rs:re])`, and on a windowed view those two disagree on 66 of 77 blocks. Every disagreement is a fetch-ahead permit taken and never returned.

The unpin/re-read spike is written and parked on `feat/unpin-and-reread` (3 of 5 tests failing, all on that divergence). It becomes a small change once #68 lands.
